core/native_test.go

```diff
@@ -0,0 +1,338 @@
+	assert.NotEqual(t, evmState.GetXShardList()[0], deposit)
+	balance, _ := shardState.GetBalance(acc1.Recipient, nil)
+	balance.GetTokenBalance(QKC)
+	assert.Equal(t, balance.GetTokenBalance(QKC), big.NewInt(10000000-21000-9000))
+	assert.Equal(t, balance.GetTokenBalance(QETHXX), big.NewInt(999999-888888))
+	assert.Equal(t, evmState.GetGasUsed(), big.NewInt(21000))
+}
+
+func TestXshardNativeTokenReceived(t *testing.T) {
+	QETHXX := common.TokenIDEncode("QETHXX")
+	QKC := common.TokenIDEncode("QKC")
+	id1, _ := account.CreatRandomIdentity()
+	acc1 := account.CreatAddressFromIdentity(id1, 0)
+	acc2 := account.CreatAddressFromIdentity(id1, 16)
+	acc3 := account.CreatEmptyAddress(0)
+	testGenesisMinorTokenBalance["QETHXX"] = big.NewInt(999999)
+	testGenesisMinorTokenBalance["QKC"] = big.NewInt(10000000)
+	env0 := getTestEnv(&acc1, nil, nil, nil, nil, nil)
+	shardSize := uint32(64)
+	env1 := getTestEnv(&acc1, nil, nil, &shardSize, nil, nil)
+	shardState0 := createDefaultShardState(env0, nil, nil, nil, nil)
+	shardState1 := createDefaultShardState(env1, nil, nil, nil, nil)
+	rootBlock := shardState0.GetRootTip().CreateBlockToAppend(nil, nil, nil, nil, nil)
+	rootBlock.AddMinorBlockHeader(shardState0.CurrentBlock().Header())
+	rootBlock.AddMinorBlockHeader(shardState1.CurrentBlock().Header())
+	rootBlock.Finalize(nil, nil, common.EmptyHash)
+	shardState0.AddRootBlock(rootBlock)
+	shardState1.AddRootBlock(rootBlock)
+	b0, _ := shardState0.CreateBlockToMine(nil, nil, nil, nil, nil)
+	shardState0.FinalizeAndAddBlock(b0)
+	b1 := shardState1.CurrentBlock().CreateBlockToAppend(nil, nil, nil, nil, nil, nil, nil, nil, nil)
+	b1.Header().ParentHash = rootBlock.Header().Hash()
+	val := big.NewInt(888888)
+	gas := new(big.Int).Add(big.NewInt(9000), big.NewInt(21000)).Uint64()
+	tx := createTransferTransaction(shardState1, id1.GetKey().Bytes(), acc2, acc1, val, &gas, nil, nil, nil, &QKC, &QETHXX)
+	b1.AddTx(tx)
+	deposit := types.CrossShardTransactionDeposit{TxHash: tx.Hash(), From: acc1, To: acc2, Value: &serialize.Uint256{Value: val}, GasPrice: &serialize.Uint256{Value: big.NewInt(1)}, GasTokenID: QKC, TransferTokenID: QETHXX}
+	txL := make([]*types.CrossShardTransactionDeposit, 0)
+	txL = append(txL, &deposit)
+	txList := types.CrossShardTransactionDepositList{TXList: txL}
+	shardState0.AddCrossShardTxListByMinorBlockHash(b1.Header().Hash(), txList)
+	rootBlock = shardState0.GetRootTip().CreateBlockToAppend(nil, nil, nil, nil, nil)
+	rootBlock.AddMinorBlockHeader(shardState0.CurrentBlock().Header())
+	rootBlock.AddMinorBlockHeader(shardState1.CurrentBlock().Header())
+	rootBlock.Finalize(nil, nil, common.EmptyHash)
+	shardState0.AddRootBlock(rootBlock)
+	b2, _ := shardState0.CreateBlockToMine(nil, &acc3, nil, nil, nil)
+	shardState0.FinalizeAndAddBlock(b2)
+	balance1, _ := shardState0.GetBalance(acc1.Recipient, nil)
+	assert.Equal(t, balance1.GetTokenBalance(QETHXX), new(big.Int).Add(big.NewInt(999999), big.NewInt(888888)))
```

Comment:
not equal

Reply:
https://github.com/QuarkChain/pyquarkchain/blob/08aa932343780279254cdd30dd665a090f931d46/quarkchain/cluster/tests/test_native_token.py#L390